### ISSUE TITLE
[MLIR][Brainstorm] Make MLIR become part of release MIOpen package

### ISSUE THREAD

**Issue by jerryyin** [ROCmSoftwarePlatform/MIOpen#919]
## Existing State

MLIR is an experimental compilation option for MIOpen. Assuming that the MLIR library and header has been shipped at ROCm release folder, a developer can use `-DMIOPEN_USE_MLIR=On` to turn on MLIR binary backend. This ticket is the umbrella item that documents what needs to be done to make MIOpen release package to also take advantage of MLIR capabilities.

### June Demo

All MLIR effort has been geared towards June demo, such that one can run `resnet50` NHWC model in TensorFlow using MLIR supported solvers in MIOpen. This means that by June demo we support:
 - MLIR as an experimental option of MIOpen: `-DMIOPEN_USE_MLIR`
 - MLIR as an optional test option of MIOpen: `-DMIOPEN_TEST_MLIR`
 - Hip backend, since framework supports `miopen-hip` only
 - Package delivery use cmake build
 - Find API. MLIR solvers won't work in immediate mode because `GetWti()` not implemented
 - No tuning for MLIR solvers. ~~In general we do not have tuning capability for `NHWC`~~
 - Perfdb dependency. We consider MLIR to be part of MIOpen, not as an independent release package. Also to achieve the best performance, MLIR need to have access to perfdb.
 - ~`GEMM` explicitly disabled for `WrW`. Gemm by default run in `NCHW`, which give wrong result for `NHWC`.~
 - Is not compatible with COMGR, since both ship with llvm and symbol collision may exist.
 - MLIR shipped as static library
 - MLIR capable of static shape only. No dynamic kernel supported but they can/will be supported
 - MLIR exposed binary interface to MIOpen
 - Revisit existing MIOpen assumptions. 

## Work items to make MLIR part of released MIOpen

We have to revisit each of the above item one by one to decide the complete work needed.

#### MLIR as an experimental option of MIOpen

For this we need to flip the switch. We either get rid of `-DMIOPEN_USE_MLIR`, or keep the option, but make release build with the option. The further implication of this is such that MLIR become part of ROCm release. 

#### MLIR as an optional test option of MIOpen

We need to do the same thing as above section for `-DMIOPEN_TEST_MLIR`. Besides, instead of making only smoke stages test MLIR, we need to make full test stages to also pick up MLIR for the guarantee of MLIR-based solver correctness.

#### Hip backend

MIOpen supports both opencl and hip backend. We need to also make sure opencl backend works, and get tracked in CI.

#### Package delivery

MLIR package delivered to MIOpen through building libMLIRMIOpen.a static library. From MIOpen perspective, this needs to be amended to using cget. 

~~An orthogonal topic, from MLIR perspective, there needs to be a way to deliver the static library and its header together with ROCm release.~~

#### Find API

We need to decide how MLIR solvers behave in non-find mode. 

#### No tuning for MLIR solvers

There are two orthogonal topics here. 

1. ~~MIOpen doesn't support NHWC tuning in general. Since part of MLIR solvers capability is the support of NHWC, MIOpen doesn't support the tuning for that partial of solvers.~~
2. Tuning support feature in MLIR solvers. By far MLIR is just a kernel generator, we need dedicated APIs for tuning. This means additional work in MLIR, MIOpen and the interface between them. However, we may want to first decide if we want tuning at all for MLIR solvers. If the performance impact between non-tuned and tuned solvers are not as much, we may not need this feature.

#### Perfdb dependency

Once tuning functionality is implemented, both MIOpen and MLIR will be dependent on perfdb. Right now if MIOpen release package come with MLIR, this won't be a problem. In the future when MLIR is released as an independent package (and no MIOpen is installed), MLIR will be able to work in only heuristic mode.

#### ~~`GEMM` explicitly disabled for `WrW`~~

~~This blocks MIOpen to populate correct result for NHWC format. We want to make `GEMM` to be solvers, for all 3 directions, `WrW` included.~~

#### Is not compatible with COMGR

We have two options:

1. Make MLIR based on the same commit of LLVM as COMGR does. This eliminate the possibility of symbol conflict.
2. Decide whether it make sense to build the two features together. It may be okay to leave COMGR and MLIR as two independent features.

#### MLIR shipped as static library

We may choose MLIR to be shipped as dynamic library. This will go against embedded builds. However, again, we want to decide on this point and this might be okay.

#### MLIR capable of static shape only

What impact can MLIR has to MIOpen when MLIR also support dynamic shape? Does it change any MIOpen existing assumptions?

#### MLIR exposed binary interface to MIOpen

Decide on the best and most robust kernel binary interface. Being able to hide the `StridedMemRef5D` make it a more robust interface. We want to revisit the alternatives and use cases to make this decision.

#### Revist existing MIOpen assumptions

- If 3x3 kernels, in NCHW mode, MIOpen by default fall back to Winograd. We may want to double check if Winograd is better than MLIR kernels.
- ~~Immediate Mode by default fall back to GEMM. However GEMM supports NCHW only. MLIR kernels may be the option to fall-back to if it support wide enough configs, NCHW and NHWC.~~
- ~~Pre-compiled kernels. It doesn't make sense for MLIR kernels to be pre-compiled. This is because MLIR kernels compilation time is trivial and it doesn't worth it to trade space with runtime for MLIR kernels. How does it impact MIOpen behavior if MLIR kernels does not get pre-compiled?~~ Nothing needs to be done on pre-compiled kernels end.

## Summary of features:

| Features                                                     | Effort |
| ------------------------------------------------------------ | ------ |
| MLIR as a build option of released MIOpen                    | Small  |
| MLIR as a test option of released MIOpen                     | Small  |
| MLIR solvers support opencl backend                          | Medium      |
| MLIR solvers support immediate mode                          | Small  |
| Tunning support for MLIR solvers                             | Large  |
| MLIR package deliver through cget                            | Small  |
| MLIR multi-threaded library compilation tests                | Medium |
| Misc features (env var to control MLIR solver, 3x3 winograd NCHW fallback, Deprecate cpp backend) | Medium |

**Comment by atamazov** [comment]
> Normal find mode. MLIR solvers won't work in immediate mode because all MLIR kernels are populated in run time

MLIR Solver uses the same machinery as all other MIOpen's Solvers for populating kernels. So the "kernels are populated in run time" cannot be a reason for a Solver to be unused in Immediate mode. The library is able to build any kernel (including MLIR) when necessary.

The reason why the library do not use some Solver via Immediate Mode API is that the library does not "know" if a Solver is applicable and how performant it is. The library can take information about applicability and performance from the find-db. If find-db record is missing, then the library checks `GetWti()` from all applicable Solvers. 

If a Solver is used in Immediate Mode API for some problem config depends on:
- (A) If a Solver is applicable for that problem config (obvious condition),
- (B) If the problem config is one of those we usually pre-configure the System-find-db for (during the "tuning" process), AND if a Solver was enabled during tuning.
- (E) If the problem config was previously run on the same system via the Find API (that updates the User-find-db).
- (D) If a Solver provides a `GetWti()` method that returns value > 0,

Please let me know if you want MLIR Solvers to be usable via Immediate Mode API during June demo.

**Comment by atamazov** [comment]
> GEMM explicitly disabled for WrW. Gemm by default run in NCHW, which give wrong result for NHWC.

This is already fixed.

**Comment by atamazov** [comment]
> "Normal find mode"

I highly recommend rephrasing use the "Find API" term instead. MLIR solvers are supported in all Find Modes, including Fast Find mode, Hybrid Find mode, Dynamic Hybrid Find mode.

**Comment by atamazov** [comment]
> MIOpen doesn't support NHWC tuning in general. 

This is supported.

**Comment by atamazov** [comment]
> Pre-compiled kernels. It doesn't make sense for MLIR kernels to be pre-compiled.

Right now the generally adopted assumption in MIopen team is that even assembly kernels should be pre-compiled.

> How does it impact MIOpen behavior if MLIR kernels does not get pre-compiled?

My expectation is some very small performance drop of Initial Iteration Time.

**Comment by mvermeulen** [comment]
One comment on tuning costs.

I expect MLIR to be used in multiple contexts, e.g. (1) generating code for MIOpen kernels (2) supporting TF/XLA or similar framework and (3) integration with MIGraphX.  Not all these are fully online or fully offline but potentially a mix of both.  In the context of some of the more offline scenarios e.g. MIGraphX "model optimizer" followed later by deployment - we may want some tuning and be willing to pay some of the costs since it is separated from the execution stage later.

So I think some form of tuning will still be useful for these sorts of scenarios.

**Comment by jerryyin** [comment]
I synced with @atamazov offline and below are summary of our conversations:

#### Tuning:

Key concerns comes from the following area:
 - Perfdb access time for the initial round can be as high as 200 ms. However, this is not a problem for frameworks because that is done only once.
 - New kernels will focus more on dynamic kernels. Dynamic kernels does not take advantage of perfdb good enough. - We need to have independent entries for each config that dynamic kernels tuned against.
 - Tuning (maintenance) overhead is quite high.

The key question therefore becomes a way to evaluate igemm kernels fairly. We ended up doing this in #939. 

As it turns out in #939, we can still achieve 5x speedup in tuning in best case scenario, and on average 3x speedup in WrW tuning. Considering this, tuning feature is still necessary. It is possible that disable solvers' whose tuning is not as effective after the tuning feature is implemented.

#### COMGR:

COMGR and MLIR continue to be independent features. Unless required by customers, we will continue to keep them independent.

#### Static library

MLIR can continue to ship as a static library and get statically linked to MIOpen. The only caveat is that customer with LLVM symbols will not be able to statically link MIOpen, but use MIOpen as a dynamic library. MLIR can ship as a dynamic library should we have the requirement, we just don't have the incentive to ship MLIR as one dynamic library now.

#### Existing binary interface

From MIOpen's perspective, the existing binary interface is simple and performant enough.

From MLIR's perspective, the interface is still quickly evolving, we should not optimize this field unless it is completely stablized.

**Comment by pfultz2** [comment]
> MLIR can continue to ship as a static library and get statically linked to MIOpen.

So we would disable MLIR when building MIOpen statically?

> The only caveat is that customer with LLVM symbols will not be able to statically link MIOpen, but use MIOpen as a dynamic library.

Statically linking HIP will bring in LLVM symbols.

> MLIR can ship as a dynamic library should we have the requirement, we just don't have the incentive to ship MLIR as one dynamic library now.

But MLIR still needs to be a static library as users wanting to build MIOpen as a static library want to build everything statically. More importantly, some frameworks want to statically link rocm in their pip wheel, so this type of deployment may become more common.

**Comment by pfultz2** [comment]
> COMGR and MLIR continue to be independent features. Unless required by customers, we will continue to keep them independent.

We should be enabling COMGR by default, as this was a requirement for pytorch's pip wheel. So I dont see how we can treat it as independent. They both need to work together.

**Comment by jerryyin** [comment]
@pfultz2 Thanks for bring these up.

> So we would disable MLIR when building MIOpen statically?

If COMGR has to come with MIOpen's static release package, then yes. I don't see any requirement that MLIR must be available in that scenario. 

My understand is that this milestone can target MLIR to be built as part of the dynamic release package of MIOpen. 

@atamazov Please correct me if I'm wrong.

> We should be enabling COMGR by default, as this was a requirement for pytorch's pip wheel. 

I might need more clarification on it. Who is the stakeholder on this front? I will ping Jeff Daily on it. I need to clarify whether MLIR is a required feature under that scenario.

**Comment by pfultz2** [comment]
> > We should be enabling COMGR by default, as this was a requirement for pytorch's pip wheel.
> I might need more clarification on it.

Well, better said, MIOpen needs to be able to compile programs without calling to `clang` or using headers on the filesystem of pytorch's pip wheel. This also needs to be done more widely as we move to "no path" instead of using absolute or relative path. 

> However, I seriously doubt it that's achievable considering everything is in so fast development stage. 

Ideally, we would move MLIR into our rocm fork. This way the team that merges with LLVM upstream can take care of breaking changes(so we aren't duplicating efforts).

Alternatively, can't we rename symbols(`-Dllvm=rocm_lllvm_mlir`) to avoid conflicts?

**Comment by jerryyin** [comment]
> Ideally, we would move MLIR into our rocm fork. 

This we are on the same page. I just don't know if MLIR is at that maturity level yet.

> Alternatively, can't we rename symbols(-Dllvm=rocm_lllvm_mlir) to avoid conflicts?

Refer to [here](https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/80) for everything I have tried the previous time. That `-D` approach doesn't work because dumped symbols does not change at all before and after applying this. Also BTW, llvm namespace is just one of the many namespaces... Internally we also enabled `mlir`, `lld` and so forth. I don't see how this redefinition of namespace can work or scale.

I'm also reading through a number of materials, including [this stackoverflow answer](https://stackoverflow.com/questions/12862975/statically-link-a-private-library-into-a-public-one-to-hide-symbols). My conclusion has been basically that to hiding or renaming or internal linking those symbols are just not possible. Unless we are on ROCm release and using the same commit of LLVM as the ROCm release.

**Comment by jerryyin** [comment]
This is to document that in the meeting @atamazov pointed out that there are duplication between:
 - Cpp igemm solvers
 - MLIR backend igemm solvers

Since Cpp igemm solvers compilation speed of slower than that from the MLIR backend, we want to disable / (de-prioritize) cpp igemm solvers.

We will postpone this feature because:
 - The MLIR backend will be only available in shared library, therefore limited scenarios. This will create a complication that we need to invent the on/off mechanism of those two groups of solvers, and make sure things like perfdb are in sync.
   - This will no longer be a problem once MLIR backend becomes available in all use cases. If we postpone this feature to that time, the on/off mechanism will no longer be needed
 - At this milestone, we still use the cpp igemm solvers as the benchmarking baseline (with perfdb on).

**Comment by junliume** [comment]
@jerryyin should I convert this issue into "discussion"?

**Comment by jerryyin** [comment]
@junliume 90% of features in this ticket are done in the past year. I'd treat this as a specification in the past. We can close this one now.

**Comment by atamazov** [comment]
@jerryyin Specifications should not be closed unless outdated/invalid. But we can assume this ticket is an enhancement. Can you please change labels? Thanks.

**Comment by jerryyin** [comment]
@atamazov Sounds reasonable either way, we can:
 - Reopen this ticket as specification as this documents the level of support between rocMLIR/MIOpen integration
 - Or, keep it closed but add the `enhancement` label treating this as outdated as we start to phase out the integration

Are you indicating the 2nd alternative?

**Comment by atamazov** [comment]
@jerryyin Yes, the 2nd one.
